### DISCUSSION TITLE
53-bit SessionID collision risk, is it a problem?

### DISCUSSION THREAD

**Discussion by cassibel** [streamich/json-joy#680] [Q&A] [answered]
Hi, this library looks fantastic, but one thing holding me back is the SessionID size. 

Why not use a full 128-bit UUID? Unless there is a central server coordinating everything, it seems possible for two forks of a document to end up with two identical SessionIDs that then cannot be later merged.

**Comment by streamich** [comment, marked as answer]
Hi @tsukkiren, the 53-bit session IDs (`sid`) should not be a problem. I think even smaller values would work.

The reason for choosing 53-bit values is because it is the largest numeric value in JavaScript with which one can performantly work with. Storing and working with 128-bit UUID is inefficient. There are storage costs, but also runtime costs: all algorithms need to constantly do `sid` comparisons and having a `sid` in a `number` is the most efficient way.

Also, runtime memory costs, all document nodes and RGA chunks have an ID assigned to them, when a document with thousands of nodes is loaded, it is thousands of nodes times the number of bytes the `sid` consumes. If the `sid` is represented by UUID I reckon it woud use at least 10x more memory (be they stored in `Uint8Array` or `[number, number, number, number]` 4-tuple).

The likelihood that two `sid`s can collide is one in `2 ** 53 = 9007199254740992`, even if they collide it does not necessarily mean that the document gets into a corrupted state. I have not tested it, but in theory it should get corrupted only if two peers generate the same `sid` AND edit the same part of the document (conflicting change) AND do it concurrently AND the changes are semantically different.

If there is a central server, non of that applies; the server just enforces the `Patch` order and contents. The library even has the ability to run the JSON CRDT in the "server-clock" mode, where `sid` is set to `1` for all peers.

Going forward the 53-bit value could potentially be increased to 64 bits. The 53-bit limitation currently exists in `binary` storage format (not in `compact` nor in `verbose`). The `binary` format would need to be modified to allow 64-bit values. Also, at runtime the `sid` is represented as `number`, we could change it to `bigint` in the future.

**Reply by streamich** [reply]
I think the 53-bit values should work just fine, with no changes needed for a long time.

In the future, once `bigint` support is better in JavaScript, or if 53-bit values turn out to be a problem the `sid` could be upgraded from `number` to `bigint` and the `binary` codec could be adapted to support the 64-bit `bigint`s.

---

With regards to 128-bit UUID, above I imagined that they would consume 10x more memory. As normally you would store UUID in `Uint8Array` which has something like 50 byte overhead, just for the `Uint8Arary` header, plus, 16 bytes of data. Alternatively, storing it as a tuple `[number, number, number, number]` of four 32-bit `number`s also has some `Array` overhead, plus each `number` would be 32 to 64+ bits in JavaScript.

However, there is a better way to store UUIDs. In storage codecs we would store simply as two 64-bit numbers. In JavaScript runtime we would expand the `ITimestampStruct` to "simply" store one more `number` or `bigint`, which would make the struct ~50% larger:

```js
// from:
ITimestampStruct { sid: number, time: number }

// to:
ITimestampStruct { sid: number, sid2: number, time: number }
```

**Reply by cassibel** [reply]
Hi @streamich , Thanks for your thoughtful reply! 

> The likelihood that two `sid`s can collide is one in `2 ** 53 = 9007199254740992`

This is true if there are only two users editing a document. With a few million users (or rather devices & browser tabs) at 53-bits you'd have a very realistic collision risk. 

I guess you're imagining a Google Docs kind of use case with a handful of users concurrently editing a document on a centralized server. In this case I agree 53-bits is fine. 

I'm imagining a web of documents that can be passed around and freely forked offline, where any two forks should be able to later merge back together. In this case, at scale, collision risk is a problem. (Especially, e.g., if you keep a map of `sids` to user id and want to be able to attribute any given change to a given user, or want to be able to move a section from one document to another while keeping such attribution. Then `sid` best be unique across all documents).

In terms of storage efficency, I imagine you could index UUID `sid`s at the start of a patch file, and simply refer to `sid` by index elsewhere in the patch. Would be a lot smaller than 53-bits. I guess same for memory efficiency while processing patches (mapping from patch `sid` index to process `sid` index).

**Reply by streamich** [reply]
For a Google Docs-like scenario all this is not a problem at all. As mentioned above, for a centralized server scenario, the library can use timestamps (`ServerTimestamp`) where `sid` of all peers is set to `1` and it works out fine as the central server can re-stamp all the operations (it also has to enforce idempotency). So for centralized scenarios the `sid` can be 0-bits.

For a non-centralized scenario, I actually have very much in mind a local-first web4 use case, where users share their documents in low- or no-bandwidth high-latency networks. (Say, your collaborator is on Mars.)

Browser tabs should re-use the same `sid`. Have a shared document, and perform cross-tab resource locking.

I don't think there will ever be millions of contributors for a single document. Simply because all session IDs need to be stored in the *clock table*. Say, you had a million contributors each identified by their own `sid`, which is UUID (16 bytes), that is 16 MB of data just to store the clock table.

It is impractical to store 16 MB of metadata, just for the clock table.

And not really possible to get into this state either: say 10 new users (devices) edit that document every day. It would take 100,000 days to reach 1 million unique `sid`s. (100,000 days is 273 years.) Linux is 30 years old and has about 30K contributors. At this rate, it would take Linux project 1,000 years to get to 1 million contributors.

Even if a document had 1 million unique contributors the collision chance would still be low: `2 ** 53 / 1000000 = 9007199254`. (This actually can be improved, no? For "small" user counts (less than a billion users): when picking a random `sid` user could check if that `sid` was already taken, then the probability would approximately not change regardless of the number of users, it would be `2 ** 53 - 1000000` instead of `2 ** 53 / 1000000`.) And, as mentioned above, if `sid`s collide it still is not a problem, unless users&mdash;with the same `sid`&mdash;edit the document at the same logical time&mdash;make concurrent changes. Basically, you can actually reuse `sid`s after some time.

In a high-latency scenario&mdash;say IPFS or peer-to-peer network&mdash;there will simply not be 1,000,000 unique `sid`s. (Nor actually, you need to enforce `sid` uniqueness if your data structure in IPFS is a Merkle DAG, but that is its own discussion.)

In a low-latency scenario there are actually almost no concurrent changes happening. (It is known from practice, people have been editing text documents collaboratively. Concurrent changes happen but are rare.) So even if users have the same `sid`, their changes will merge just fine as they will not be concurrent.

In zero-latency scenario, CRDTs and all this tooling is actually not necessary, but will work just fine. As there will be no concurrent changes by definition.

The most challenging scenario&mdash;and I am explicitly building this library exactly for that scenario&mdash;is when you have a handful of users with high-latency editing the same document. Say, a group of users are hiking in mountains, another group of users are sailing on a boat, some user is incarcerated and gets access to internet once a day, and few other users are on Moon and Mars. They all edit this one document concurrently but changes are synced only every 20 min to 4 days, hence, there are always lots of conflicts. The library is tuned to handle this constant high-conflict scenario. In this case you would have up to few hundred `sid`s and the 53-bit `sid` space should be enough. Hopefully. 🤞

If it is not enough, we later extend it to 64-bit or 2x 53-bit, or 2x 64-bit `sid`s.

The `sid` is not intended to be used for attributing changes to the user. There are `Patch` objects which represent changes, if the `Patch` needs to be attributed to some user: there is a custom optional *metadata* section where any JSON can be stored in the `Patch`, one can store user ID there.

For storage, `sid`s are already stored as a *clock table* in the document and are referenced by index (in `binary` codec).

**Reply by cassibel** [reply]
Your math is off on the collision chance calculation (it's about 1/180 for a million 53-bit ids, see the Birthday Problem), but otherwise what you say makes a lot of sense.

What I'm thinking in addition to what I already said is that JSON documents usually represent some kind of structured data, where each JSON object is likely to need an id (like a you add an id to every row in a database, so objects can be identified and reference other objects). If every object needs its own id anyway, and we're working offline so we can't generate ids serverside, then it makes a lot of sense to reuse the logical timestamp as an object id. And if ids need to be unique across all documents then they should be ~128-bits. 

The alternative would be to keep sids at 53-bit for patches and generate a second logical timestamp with a 128-bit sid for each object id, which obviously takes up more space. 

Given there's a clock table so each `sid` is only stored once in any document or patch, it seems to me a small cost to bump `sid`s up to 128-bits so timestamps can be reused as object ids.

**Reply by cassibel** [reply]
Question: Is there actually any cost to increasing the sid size to 128 bits? It seems it would cost an extra 64 bits per patch in storage, and an extra 64 bits per id in a document. I guess that's insignificant even with thousands of patches and dozens of users?

Would it slow the patch processing time at all? If an sid is just a 64-bit pointer to a 128-bit value in memory then it seems no more memory cost per patch item than directly using a 64-bit sid value.

**Reply by cassibel** [reply]
Also

> I don't think there will ever be millions of contributors for a single document. Simply because all session IDs need to be stored in the clock table. Say, you had a million contributors each identified by their own sid, which is UUID (16 bytes), that is 16 MB of data just to store the clock table.

I don't mean a million contributors in a single document but a million contributors across a million forks, any of which could be merged with any other.

I agree collision risks are still low, but not sure I see the advantage to keeping at 53/64 bits when seems so cheap to bump up the size and not have to think about these things.

**Reply by streamich** [reply]
I don't think Birthday Problem applies that way if `sid` is checked for all already known used `sid`s. To use Birthday Paradox calculation here one would need to take as N the number of concurrently chosen `sid`s. So, if at the same time 3 users pick a `sid` we need to calculate `3` out of `2 ** 53` probability. (Which according to the Birthday Paradox formula makes it less likely than one in quadrillion (million billion).)

Costs of storing extra `sid` entropy are not small. Patch and model sizes would grow by more than 64 bits for each `sid` stored. Assuming there are million million documents the extra storage cost would be well into petabytes. Also, code changes and runtime performance.

We are not increasing the `sid` entropy unless there is a good concrete reason. So far I have not seen one.

It does not matter how many millions of contributors and how many millions of forks there are. What matters is how many of them are actually merged, if only two forks are merged and users of those forks picked `sid`s concurrently that is Birthday Paradox probability of `N = 2` out of `2 ** 53`.

If those users did not pick their `sid` concurrently. For example, one user forked from another. There will be 0% chance of collision, as that second user will check for all known `sid`s when forking.

I did not really follow the idea of document IDs. Document IDs should be something completely external, which internal model knows nothing about.

**Reply by cassibel** [reply]
> I did not really follow the idea of document IDs. Document IDs should be something completely external, which internal model knows nothing about.

I meant the reverse (document aware of the internal model), but on second though reusing the CRDT logical clock for entity ids in the JSON document a bad idea (if you copied an entity to another document the clock value and therefore id would change). Makes sense to keep these things separate. 

Other than that, you're right. Birthday Problem applies to only concurrent chosen `sid`s later merged into one document, which should never approach a million. I'm trying hard to find a scenario where collision risk is likely and not finding one. I think my 1/180 above is actually 1/18,000, dropping to 1 / 18billion for 1000 concurrently chosen ids. 

Maybe just change `Math.random()` to cryptographic randomness in the `sid` generator function and 53-bits is good.

Thanks for thinking this through with me.

**Reply by streamich** [reply]
Thanks for you inputs, I've updated the forking code, now when forking, out-of-the-box [the document will check for all already consumed `sid`s](https://github.com/streamich/json-joy/pull/681/commits/6d3cdf7dea10d8c87f302c9e5361f8fe7719a4ce#diff-43a082ef32a363be88ee43b8324948b9133739c5a10ce95c51ef5e976091281dR51). Which means the collision chance is 0% when you are "connected". And when you don't have network, the collision chance is the number of forks created while all peers were offline out of `9007199253740992` possibility space, as per Birthday Paradox formula (one in million billion chance for five forks created concurrently).

I don't think we need cryptographically secure random number generator here. Because we generate only one number, and it happens very infrequently on different user's devices in different moments of time; their randomly seeded `Math.random()` should be enough.

V8 uses 128 bits of entropy in `Math.random()`, we need only 53 bits.
